### ISSUE TITLE
Why does Zygote differentiate unnecessary code?

### ISSUE THREAD

**Issue by jessebett** [FluxML/Zygote.jl#571]
This discussion was started on Slack, but not resolved:

Okay, so I know I’ve asked this a few times. This is related to Zygote failing to support `@info` or string interpolation inside a function being differentiated. Though this issue has been solved with clever `@nograd` definitions, I think it indicates a general problem that I would like to understand.

Why does Zygote attempt to differentiate code that does not affect the primal value?

```
using Zygote
f(x) = 2x
losses=[]
Zygote.gradient(1.) do x
  loss = f(x)
  append!(losses,loss)
  return loss
end
```

For example, the above code triggers the `mutating arrays not supported` error. This kind of problem can't be solved like above with `nograd` because in general we might want to differentiate through mutating arrays. However, this case the primal value is unaffected by it being appended to an array, mutation should not be part of the expression being differentiated. 

My understanding is that Zygote can’t guarantee in general that functions called don’t have side effects or closures, but what I don’t understand is why that information isn’t available to be inferred by Zygote’s IR in this case (or the case where we’re logging or printing).

Specifically this should not require any ability to mutate arrays, it only requires realizing that the returned value isn’t affected by that array mutation. Same with differentiating `@info`, shouldn’t involve anything with `try/catch`. That this is being solved by `nograd` on specific functions is fine for printing and logging, but what is preventing Zygote from inferring what code is actually affecting the primal value?

@oxinabox @MikeInnes can you shed some light for me why Zygote can't infer the information required to address this more generally? (I pinged on Slack already, but here as well) @

This is quite a common pattern. e.g logging or appending the primal value of a loss function during training. Also that something like `@nograd` can’t be used inside the do block to manually tell Zygote to ignore it. @oxinabox can you link the PR for the function you were describing that could be used inside the gradient like `@ignore`?

```
Zygote.gradient(1.) do x
  loss = f(x)
  Zygote.@ignore append!(losses,loss)
  return loss
end
```

**Comment by jessebett** [comment]
For example, using Zygote's IR:

```
vals = []
f(x) = 2x

function g(x)
  y = 2. * x
  append!(vals,y)
  return y
end

Zygote.@code_ir g(1.)
```

We get:

```
julia> Zygote.@code_ir g(1.)
1: (%1, %2)
  %3 = 2.0 * %2
  %4 = Main.append!(Main.vals, %3)
  return %3
```

Since the function `g` returns `%3`, why should `Zygote.gradient` touch anything after `%3`?

```
julia> Zygote.gradient(g,1.)
ERROR: Mutating arrays is not supported
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] (::Zygote.var"#999#1000")(::Nothing) at /Users/jesse/.julia/packages/Zygote/oMScO/src/lib/array.jl:49
 [3] (::Zygote.var"#2658#back#1001"{Zygote.var"#999#1000"})(::Nothing) at /Users/jesse/.julia/packages/ZygoteRules/6nssF/src/adjoint.jl:49
 [4] _append! at ./array.jl:923 [inlined]
 [5] (::typeof(∂(_append!)))(::Nothing) at /Users/jesse/.julia/packages/Zygote/oMScO/src/compiler/interface2.jl:0
 [6] append! at ./array.jl:915 [inlined]
 [7] g at ./REPL[3]:3 [inlined]
 [8] (::typeof(∂(g)))(::Float64) at /Users/jesse/.julia/packages/Zygote/oMScO/src/compiler/interface2.jl:0
 [9] (::Zygote.var"#28#29"{typeof(∂(g))})(::Float64) at /Users/jesse/.julia/packages/Zygote/oMScO/src/compiler/interface.jl:38
 [10] gradient(::Function, ::Float64) at /Users/jesse/.julia/packages/Zygote/oMScO/src/compiler/interface.jl:47
 [11] top-level scope at REPL[28]:1
```

**Comment by jessebett** [comment]
@oxinabox talked a bit about mutation being tricky. Here's an example of why:

```
function g(x)
  y = [2. * x]
  append!(y,y)
  return y
end

julia> Zygote.@code_ir g(1.)
1: (%1, %2)
  %3 = 2.0 * %2
  %4 = Base.vect(%3)
  %5 = Main.append!(%4, %4)
  return %4
```

If we sliced out the expressions that don't affect the return `%4` then in the case where we have mutating code `%5` is not captured.
